crates/runtime/README.md

```diff
@@ -0,0 +1,662 @@
+When the execution of a Wasm function concludes, `VMLogic` is responsible for
+gathering all the results, logs, and events, and packaging them into an
+`Outcome` struct. This struct provides a comprehensive summary of the execution,
+including any return values or errors encountered during the process. By
+centralising this information, `VMLogic` ensures that the caller of the runtime
+receives a complete and consistent report of the execution results.
+
+### Host functions
+
+Host functions are implemented in the `imports.rs` file:
+
+  - Panic handling and UTF-8 string management
+  - Register operations (read, write, get length)
+  - Storage interactions
+  - Logging and event emission
+  - External HTTP requests (fetch)
+
+Host functions serve as the bridge between the sandboxed WebAssembly (Wasm)
+environment and the host system. These functions provide a carefully curated set
+of capabilities to the guest Wasm code, allowing it to interact with the outside
+world in a controlled and secure manner. The primary purpose of host functions
+is to extend the functionality of the Wasm module beyond its inherent
+limitations while maintaining strict control over its access to system resources
+and external services.
+
+At the core of the host function implementation is the `VMHostFunctions` struct,
+which encapsulates the runtime's logic and provides a safe interface for the
+guest code to interact with. This struct is created using self-references,
+allowing it to maintain a consistent view of the execution environment
+throughout the lifetime of a function call.
+
+One of the fundamental responsibilities of host functions is memory management.
+They provide methods for reading from and writing to the guest's memory space,
+ensuring that all memory accesses are bounds-checked and secure. This includes
+functions for reading raw bytes from memory as well as higher-level operations
+like retrieving UTF-8 strings. By mediating all memory interactions, the host
+functions prevent potential security vulnerabilities that could arise from
+unchecked memory access.
+
+Register operations form another critical aspect of the host functions. They
+allow the guest code to store and retrieve data in a temporary storage system,
+which is particularly useful for passing information between different parts of
+the Wasm module or for storing intermediate results. The host functions ensure
+that all register operations adhere to the size and count limits specified in
+`VMLimits`, preventing resource abuse and maintaining the integrity of the
+execution environment.
+
+Storage interactions are also facilitated through host functions. These
+functions provide a key-value storage interface, allowing the guest code to read
+from and write to a persistent storage system. By channeling all storage
+operations through these functions, the runtime can enforce access controls,
+manage resource usage, and potentially implement advanced features like rollback
+or versioning.
```

Comment:
```suggestion
from and write to a persistent storage system.
```

- access-control (I guess in the context of contexts not reading other context data) isn't managed by the runtime, but by the node
- resource usage (I guess storage usage) for now isn't metered nor is it constrained
- rollback and versioning would also fall outside the purview of the runtime